Feature/file upload

Sam09040 opened the pull request:
- added email service;
- added csv service;
- added upload to graphql;
- added createMany use case;

Sam09040 committed:
feat: file upload and email

- added graphql-upload-ts for file upload;
- using resend to send emails;
- added middleware to server;
- added crypto service for random password generation;
- added insertMany for user and address;
- created .csv files for testing;

Sam09040 committed:
post rebase fixes

- most are changes made after running prettier;

Sam09040 committed:
upload tests

- added tests for upload use case;

Sam09040 committed:
fixes

- added instructions to readme;
- fixed the create many test;
- updated migrations to change the tables' names;

walteraandrade commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,82 @@
+import { UserDbDataSource } from '@data/user';
+import { isDefined } from '@test/utils/is-defined';
+
+describe.only('UserResolver - CreateManyUsers', () => {
```
```suggestion
describe('UserResolver - CreateManyUsers', () => {
```

walteraandrade commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,82 @@
+    await disconnectDb();
+  });
+
+  it.only('should create a single user', async () => {
```
```suggestion
  it('should create a single user', async () => {
```

walteraandrade commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,82 @@
+    formData.append('operations', JSON.stringify(query));
+    formData.append('map', JSON.stringify({ 0: ['variables.file'] }));
+    formData.append('0', file);
+    const response = await requestMaker<any, any>({ formData });
```
any? 🤔 
não dá pra saber qual o tipo aqui?

walteraandrade commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,82 @@
+    formData.append('0', file);
+    const response = await requestMaker<any, any>({ formData });
+    expect(response.data.data).to.equal('Upload ended successfully! Check the database to see the uploaded info.');
+    const user = await datasource.findAll();
```
eu sempre prefiro que os passos de avaliação, i.e. os `expect` fiquem juntos, acho que torna a leitura do teste mais fácil

walteraandrade commented on src/api/graphql/module/user/create-user.test.ts:
```diff
@@ -1,25 +1,20 @@
       }
-    }
-  `;
+      }
```
a indentação está quebrada

walteraandrade commented on package-lock.json:
meio grande essas mudanças 👀 
você sabe por que ele veio grande assim?

Sam09040 replied:
provavelmente pelas mudanças de dependências que eu fiz na tentativa de fazer o mocha funcionar.

Sam09040 committed:
fix: tests

- changed mocha for jest because of the problems;
- removed snaplet because it required esm and conflicted with project;
- removed tsconfig-paths-dist.js;
- changes made because of change in the tsconfig configuration;

Sam09040 committed:
fix: .only

- removed .only from users test;

racsandoval commented on src/api/graphql/module/address/address.type.ts:
```diff
@@ -6,27 +6,27 @@ export class Address implements AddressModel {
   userId: number;
 
-  @Field({ description: 'Address cep' })
+  @Field(() => String, { description: 'Address cep' })
```
pq trocou a tipagem inferida das strings por uma explicita?

Sam09040 replied:
durante o processo de correção do mocha, o walter, ao utilizar a configuração do tsconfig de outro projeto, rodou os testes com o mocha e o erro trocou de ERR_UNKNOWN_FILE_EXTENSION para um erro relacionado a tipagem de uma variável, então nós achamos que essa configuração tinha consertado o problema do mocha, mas, na realidade, o erro de tipagem foi, de alguma forma, "mais rápido", no sentido de que apareceu antes, que o erro de extensão desconhecida (tenho imagens que mostram esse comportamento). dito isso, eu não alterei a configuração de volta pois essa verificação de tipagem é mais comum entre os projetos, então é bom para eu já estar ciente do que eu vou encontrar.

walteraandrade replied:
na verdade, esse erro aconteceu logo que tentei rodar os testes na develop, usando as configurações normais do projeto 🤔

Sam09040 replied:
esse erro só me ocorreu depois que eu coloquei as configurações do tsconfig que você mandou. antes, rodava normalmente para mim, sem aparecer erros mais específicos de tipagem, como uma variável 'string' não poder ser 'string | undefined'.

racsandoval commented on src/api/graphql/module/address/address.type.ts:
```diff
@@ -6,27 +6,27 @@ export class Address implements AddressModel {
   id: number;
 
-  @Field({ description: 'user id connected to the address' })
+  @Field(() => Int, { description: 'user id connected to the address' })
```
qual a diferença de tipar como `Int` e deixar a tipagem inferida?

Sam09040 replied:
ao utilizar a tipagem inferida padrão ou tipagem ID, ele utiliza Float, sendo que o tipo que desejamos é o Int, então, ao utilizar essa tipagem explícita, infere a tipagem Int ao invés de um tipo que não utilizamos.

racsandoval commented on src/api/graphql/module/address/create-address.test.ts:
```diff
@@ -1,13 +1,11 @@
-import { connectDb, connectServer } from '@test/utils/connect.util';
-import { disconnectServer, disconnectDb } from '@test/utils/disconnect.util';
-import { UserDbDataSource } from '@data/user';
+import 'reflect-metadata';
```
pq tem esse import aqui? ele não deveria ser utilizado apenas no primeiro arquivo lido pelo projeto?

Sam09040 replied:
eu rodei alguns testes com o jest e percebi que era necessário o uso do reflect metadata em cada arquivo de teste, senão o teste não conseguiria usar das funções do typeDI.

racsandoval replied:
por padrão esse import deveria ser feito apenas uma vez no arquivo de entrada da aplicação, caso o jest leia todos os arquivos separadamente faz sentido, mas ainda sim poderia ter um arquivo de entrada

racsandoval commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,99 @@
+    formData.append('map', JSON.stringify({ 0: ['variables.file'] }));
+    formData.append('0', file);
+    const response = await requestMaker<{ uploadCsv: string }, undefined>({ formData });
+    const user = await datasource.findAll();
```
faltou checar se o usuário foi criado corretamente

racsandoval commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,99 @@
+    formData.append('map', JSON.stringify({ 0: ['variables.file'] }));
+    formData.append('0', file);
+    const response = await requestMaker<{ uploadCsv: string }, undefined>({ formData });
+    const user = await datasource.findAll();
```
faltou validar se os usuários foram criados corretamente

racsandoval commented on src/api/graphql/module/user/create-user.test.ts:
```diff
@@ -1,45 +1,38 @@
-import { connectDb, connectServer } from '@test/utils/connect.util';
-import { disconnectDb, disconnectServer } from '@test/utils/disconnect.util';
-import { getToken } from '@test/utils/get-token.util';
+import 'reflect-metadata';
```
mesma coisa aqui

racsandoval commented on src/api/graphql/module/user/user.input.ts:
```diff
@@ -1,27 +1,27 @@
-  @IsISO8601(undefined, { message: 'Birth date must follow this pattern: dd-MM-yyyy' })
+  @Field(() => String, { description: 'Birth date' })
+  @IsNotEmpty({ message: 'The birth date must not be empty and must have the format dd-MM-yyyy' })
+  @Length(10, 10, { message: 'The birth date must have 10 characters and must have the format dd-MM-yyyy' })
```
como precisa manter o formato específico, valeria colocar um validador de regex

racsandoval commented on src/core/security/crypto/crypto.config.ts:
```diff
@@ -0,0 +1,4 @@
+import 'dotenv/config';
```
pq esse import é feito aqui?

Sam09040 replied:
ainda não sei porque, é algo que preciso olhar mais a fundo, mas em alguns lugares que eu colocava para pegar a variável do process.env, ele não pegava, então causava erro, mas ao colocar esse import, ele funcionava. ainda vou dar uma olhada melhor nisso.

racsandoval replied:
a config do `dotenv` poderia ser feita no arquivo de entrada da aplicação

racsandoval commented on src/data/db/migrations/20241218134250_fix_names/migration.sql:
```diff
@@ -0,0 +1,47 @@
+ALTER TABLE "Address" DROP CONSTRAINT "Address_user_id_fkey";
+
+-- DropTable
+DROP TABLE "Address";
+
+-- DropTable
+DROP TABLE "User";
```
essa migration causaria perda de dados, já que ela deleta essas duas tabelas

racsandoval commented on src/domain/user/create-many.use-case.ts:
```diff
@@ -0,0 +1,135 @@
+import { Service } from 'typedi';
+
+@Service()
+export class CreateManyUseCase {
```
CreateMany o que? poderia deixar mais explicito no que essa classe cria

racsandoval commented on src/domain/user/create-many.use-case.ts:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  async validateCsvData(csvData: CsvInputModel[]): Promise<unknown[]> {
+    const errorConstraints: unknown[] = [];
```
pq essa array é `unknown`? não está colocando esse formato nela?
```
{
          user: `User: ${i + 1}`,
          property: error.property,
          constraints: error.constraints,
  }
 ```

Sam09040 replied:
não tinha me atentado a isso no momento que estava fazendo essa função. já fiz algumas alterações.

racsandoval commented on src/domain/user/create-many.use-case.ts:
```diff
@@ -0,0 +1,135 @@
+
+  async sendEmails(csvData: CsvInputModel[], passwords: string[]) {
+    for (const [i, user] of csvData.entries()) {
+      //await this.emailService.sendEmail(user.name, user.email, passwords[i]);
```
pq ta comentado aqui?

Sam09040 replied:
eu imaginei que, por estar mandando utilizando de emails não existentes, haveria algum erro ao mandar, então, para evitar tal ocorrência, optei por mandar no log, o que me garante que chegou naquele ponto e teve sucesso, apesar de não utilizar do email service. contudo, eu rodei o teste com o email service algumas vezes e ele funcionou bem.

racsandoval replied:
pensando nos testes da aplicação, seria inviável ficar modificando código para garantir que os testes passem, uma solução que temos pra isso são [stubs](https://sinonjs.org/releases/latest/stubs/)

Sam09040 committed:
fix: tests

- added user check on create many test;
- added regex to validate birth date format;
- createMany to createManyUsers;

sophia-bergamo commented on src/api/graphql/module/address/address.resolver.ts:
```diff
@@ -21,7 +21,7 @@ export class AddressResolver {
   @Mutation(() => Address, { description: 'Create a new address' })
   @Authorized()
-  async createAddress(@Arg('data') data: AddressInput, @Ctx() { userId }: ServerContext) {
+  async createAddress(@Arg('data', () => AddressInput) data: AddressInput, @Ctx() { userId }: ServerContext) {
```
pq precisou fazer assim aqui? geralmente usamos como estava antes mesmo

Sam09040 replied:
eu achei que era melhor para o type graphql entender os argumentos.

sophia-bergamo commented on src/api/graphql/module/user/create-many.test.ts:
```diff
@@ -0,0 +1,125 @@
+  });
+
+  it('should create a single user', async () => {
+    const file = fs.createReadStream('./src/test/csv/single-user.csv', 'utf-8');
```
tem um conceito que usamos aqui na instaq para padrozinar os arquivos de testes chamado AAA (Arrange, Act e Assert). Basicamente separar as responsabilidades dos métodos, constantes e outras estruturas que usamos em cada `it` dos testes e facilitar a leitura deles.

Sam09040 replied:
vou dar uma olhada nesse conceito!

sophia-bergamo commented on src/api/graphql/module/user/user.resolver.ts:
```diff
@@ -25,24 +26,30 @@ export class UserResolver {
 
   @Mutation(() => Login, { description: 'Authenticate user' })
-  login(@Arg('data') data: LoginInput) {
+  login(@Arg('data', () => LoginInput) data: LoginInput) {
```
aqui também

sophia-bergamo commented on src/api/graphql/module/user/user.resolver.ts:
```diff
@@ -25,24 +26,30 @@ export class UserResolver {
-  createUser(@Arg('data') data: UserInput, @Ctx() context: ServerContext) {
-    return this.createUserUseCase.exec(data, context.token);
+  createUser(@Arg('data', () => UserInput) data: UserInput) {
+    return this.createUserUseCase.exec(data);
```
faltou o retorno da função aqui

sophia-bergamo commented on src/data/address/address.db.datasource.ts:
```diff
@@ -11,4 +11,13 @@ export class AddressDbDataSource {
+      userId: userId?.[index],
+    }));
+
+    return await dbClient.address.createManyAndReturn({ data });
```
```suggestion
    return dbClient.address.createManyAndReturn({ data });
```

sophia-bergamo commented on src/data/address/address.db.datasource.ts:
```diff
@@ -11,4 +11,13 @@ export class AddressDbDataSource {
     return dbClient.address.findMany({ where: { userId } });
   }
+
+  async insertMany(addressInput: AddressInputModel[], userId: number[]): Promise<AddressModel[]> {
```
sempre melhor mantém um argumento só nas funções pra evitar bugs de ordem dos parâmetros no futuro

sophia-bergamo commented on src/data/user/user.db.datasource.ts:
```diff
@@ -20,6 +24,16 @@ export class UserDbDataSource {
     });
   }
 
+  findManyByEmail(emails: string[]) {
```
```suggestion
  findManyByEmails(emails: string[]) {
```

sophia-bergamo commented on src/data/user/user.db.datasource.ts:
```diff
@@ -20,6 +24,16 @@ export class UserDbDataSource {
     });
   }
 
+  findManyByEmail(emails: string[]) {
```
qual o retorno dessa função?

Sam09040 replied:
Promise<User[]>

sophia-bergamo commented on src/domain/address/address.use-case.ts:
```diff
@@ -8,25 +8,25 @@ import { Service } from 'typedi';
     }
-  
+
     return await this.addressDatasource.findAddresses(user.id);
```
```suggestion
    return this.addressDatasource.findAddresses(user.id);
```

sophia-bergamo commented on src/domain/user/create-many.use-case.ts:
```diff
@@ -0,0 +1,141 @@
+    private readonly cryptoService: CryptoService,
+  ) {}
+
+  async exec(file: FileUpload) {
```
qual o retorno dessa função?

Sam09040 replied:
Promise<void>

Sam09040 committed:
small fixes

- added typed returns to some functions;
- findManyByEmail to findManyByEmails;
- changed argument input of insertMany from addressDb to not be positional;
- added null | undefined to address' complement;
- removed await from some returns;
- removed types from Arg decorators, unless needed;

sophia-bergamo commented on src/api/graphql/module/user/user.resolver.ts:
```diff
@@ -15,34 +15,41 @@ export class UserResolver {
   @Authorized()
-  users(@Arg('pageInput') pageInput: PageInput, @Ctx() context: ServerContext) {
-    return this.usersUseCase.exec(pageInput, context.token);
+  users(@Arg('pageInput') pageInput: PageInput) {
```
faltou o retorno dessa tbm

sophia-bergamo commented on src/domain/user/users.use-case.ts:
```diff
@@ -4,21 +4,10 @@ import { Service } from 'typedi';
+  constructor(private readonly datasource: UserDbDataSource) {}
 
-  async exec (input: PageInputModel, token: string | undefined) {
+  async exec(input: PageInputModel) {
```
retorno dessa

Sam09040 committed:
fix: returns

- added types to users' return on user.resolver.ts and exec on users.use-case.ts;

Sam09040 committed:
fixes: stub and dotenv

- added sinon for emailService stub;
- removed dotenv/config import from configs and added it to graphql-server.ts;

Sam09040 committed:
update create-many

- removed comment from create-many.use-case.ts;
